File: public/rl2/js/panneau/social.js
// Connexion sociale en direct (serveur : `admin2/social.rs`).
//
// Ouverte tant que la page l'est, pour un joueur connecte. Elle tient a
// jour les amis et leur statut, le groupe et les invitations, et previent
// des nouveaux messages. Coupee, elle se reconnecte d'elle-meme, de plus en
// plus doucement.
//
// Evenements (`on`) : `change` (amis, groupe ou invitations ont bouge),
// `invitation`, `lancer`, `nouvelles`, `info`, `erreur`.

/** Attente avant de se reconnecter : 2 s, puis de plus en plus, 30 s au plus. */
const REPRISE = [2000, 4000, 8000, 15000, 30000];

export class Social {
  constructor(base) {
    this.url = `${base.replace(/^http/, 'ws')}/api/social`;
    this.moi = null;
    this.amis = new Map();
    this.groupe = null;
    this.invitations = [];
    this.ecoute = {};
    this.ws = null;
    this.actif = false;
    this.essais = 0;
    this.dernierLieu = '';
  }

  on(evt, fn) {
    (this.ecoute[evt] ||= []).push(fn);
  }

  _emet(evt, data) {
    for (const fn of this.ecoute[evt] || []) fn(data);
  }

  /** Ouvre la connexion (joueur connecte). Sans effet si elle l'est deja. */
  demarrer() {
    this.actif = true;
    if (this.ws) return;
    const ws = new WebSocket(this.url);
    this.ws = ws;
    ws.onopen = () => {
      this.essais = 0;
      if (this.dernierLieu) this.envoyer({ t: 'lieu', lieu: this.dernierLieu });
    };
    ws.onmessage = (e) => {
      try {
        this._recoit(JSON.parse(e.data));
      } catch {
        // Message illisible : on l'ignore.
      }
    };
    ws.onclose = () => {
      this.ws = null;
      if (!this.actif) return;
      const attente = REPRISE[Math.min(this.essais, REPRISE.length - 1)];
      this.essais += 1;
      setTimeout(() => this.actif && this.demarrer(), attente);
    };
  }

  /** Deconnexion du compte : on ferme et on oublie tout. */
  arreter() {
    this.actif = false;
    this.ws?.close();
    this.ws = null;
    this.amis.clear();
    this.groupe = null;
    this.invitations = [];
    this._emet('change');
  }

  envoyer(m) {
    if (this.ws?.readyState === WebSocket.OPEN) this.ws.send(JSON.stringify(m));
  }

  _recoit(m) {
    switch (m.t) {
      case 'etat':
        this.moi = m.moi;
        this.amis = new Map(m.amis.map((a) => [a.id, a]));
        this.groupe = m.groupe;
        this.invitations = m.invitations;
        this._emet('change');
        break;
      case 'amis':
        this.amis = new Map(m.amis.map((a) => [a.id, a]));
        this._emet('change');
        break;
      case 'ami':
        if (this.amis.has(m.ami.id)) this.amis.set(m.ami.id, m.ami);
        if (this.groupe) {
          this.groupe.membres = this.groupe.membres.map((x) => (x.id === m.ami.id ? m.ami : x));
        }
        this._emet('change');
        break;
      case 'groupe':
        this.groupe = m.groupe;
        this._emet('change');
        break;
      case 'invitation':
        this.invitations = [...this.invitations.filter((j) => j.id !== m.de.id), m.de];
        this._emet('change');
        this._emet('invitation', m.de);
        break;
      default:
        // lancer, nouvelles, info, erreur
        this._emet(m.t, m);
    }
  }

  // ---------------------------------------------------------- actions ---

  /** Ou en est la page : `menu`, `solo` ou `partie`. Envoye au changement. */
  lieu(l) {
    if (l === this.dernierLieu) return;
    this.dernierLieu = l;
    this.envoyer({ t: 'lieu', lieu: l });
  }

  inviter(id) {
    this.envoyer({ t: 'inviter', a: id });
  }

  repondre(de, oui) {
    this.invitations = this.invitations.filter((j) => j.id !== de);
    this.envoyer({ t: 'repondre', de, oui });
    this._emet('change');
  }

  quitter() {
    this.envoyer({ t: 'quitter' });
  }

  exclure(id) {
    this.envoyer({ t: 'exclure', qui: id });
  }

  passerChef(id) {
    this.envoyer({ t: 'chef', qui: id });
  }

  estChef() {
    return !!this.groupe && this.groupe.chef === this.moi;
  }

  /** Le chef vient d'entrer dans le salon `code` : le groupe le suit. */
  lancer(code) {
    if (this.estChef()) this.envoyer({ t: 'lancer', salon: code });
  }

  /** Amis en ligne d'abord (en partie, puis menus, puis solo), puis les
   *  autres ; par pseudo dans chaque lot. */
  amisTries() {
    const rang = (a) => (!a.statut.en_ligne ? 3 : { partie: 0, menu: 1, solo: 2 }[a.statut.lieu] ?? 1);
    return [...this.amis.values()].sort((a, b) => rang(a) - rang(b) || a.pseudo.localeCompare(b.pseudo));
  }
}

/** Statut d'un ami en quelques mots, pour l'infobulle et sa fiche. */
export function statutTexte(s) {
  if (!s?.en_ligne) return 'Hors ligne';
  if (s.lieu === 'partie') {
    if (!s.salon) return 'En partie';
    return s.prive ? `En partie privée (salon ${s.salon}) · tu peux le rejoindre` : `En partie · Occasionnel (serveur ${s.salon})`;
  }
  if (s.lieu === 'solo') return 'En ligne · joue en solo';
  return 'En ligne · dans les menus';
}

File: admin2/src/social.rs
//! Social en direct : presence des amis, groupes et invitations, nouvelles.
//!
//! Chaque page du jeu d'un joueur connecte garde une connexion ouverte
//! (`social_ws.rs`). Par elle, le serveur pousse :
//!
//!   etat         a l'arrivee : amis et leur statut, groupe, invitations
//!   ami          un ami change de statut (connexion, menu, solo, partie)
//!   amis         la liste d'amis a change (demande acceptee, rupture)
//!   groupe       son groupe a change (ou `null` : plus de groupe)
//!   invitation   un ami l'invite dans son groupe
//!   lancer       le chef lance une partie : rejoindre ce salon
//!   nouvelles    un message ou une demande attend : relire la messagerie
//!   info, erreur une phrase a montrer
//!
//! et recoit : `lieu`, `inviter`, `repondre`, `quitter`, `exclure`,
//! `chef`, `lancer`.
//!
//! Verrous : toujours la base d'abord, puis l'etat social, jamais l'inverse.

use crate::groupes::Groupes;
use crate::presence::{Canal, Presence};
use crate::{amis, base, joueurs};
use rusqlite::Connection;
use serde_json::{Value, json};
use std::sync::{Mutex, MutexGuard, OnceLock};
use std::time::{Duration, Instant};

/// Delai avant qu'un joueur deconnecte perde sa place dans son groupe : le
/// temps de recharger la page.
pub const GRACE: Duration = Duration::from_secs(30);

#[derive(Default)]
pub struct Social {
    pub presence: Presence,
    pub groupes: Groupes,
}

pub fn social() -> MutexGuard<'static, Social> {
    static S: OnceLock<Mutex<Social>> = OnceLock::new();
    S.get_or_init(Mutex::default).lock().unwrap_or_else(|e| e.into_inner())
}

// ---------------------------------------------------------- lectures ---

fn amis_de(c: &Connection, moi: i64) -> Vec<i64> {
    amis::liste(c, moi).map(|l| l.amis.into_iter().map(|j| j.id).collect()).unwrap_or_default()
}

/// Un joueur tel que la colonne le montre : pseudo, avatar, statut.
fn fiche(c: &Connection, s: &Social, id: i64) -> Value {
    match joueurs::par_id(c, id) {
        Some(j) => json!({ "id": j.id, "pseudo": j.pseudo, "avatar": j.avatar, "statut": s.presence.statut(id) }),
        None => json!({ "id": id, "pseudo": "Joueur", "avatar": "", "statut": s.presence.statut(id) }),
    }
}

fn liste_amis(c: &Connection, s: &Social, moi: i64) -> Vec<Value> {
    amis_de(c, moi).into_iter().map(|id| fiche(c, s, id)).collect()
}

fn groupe(c: &Connection, s: &Social, moi: i64) -> Value {
    match s.groupes.groupe_de(moi) {
        None => Value::Null,
        Some(g) => json!({
            "chef": g.chef,
            "membres": g.membres.iter().map(|&m| fiche(c, s, m)).collect::<Vec<_>>(),
        }),
    }
}

fn envoyer(s: &Social, a: i64, message: Value) {
    s.presence.envoyer(a, &message.to_string());
}

/// Previent que le statut de `moi` a change : ses amis, les membres de son
/// groupe (pas forcement ses amis), et ses propres pages, qui le montrent
/// dans la zone du groupe.
fn annoncer(c: &Connection, s: &Social, moi: i64) {
    let f = fiche(c, s, moi);
    let mut qui = amis_de(c, moi);
    qui.extend(s.groupes.coequipiers(moi));
    qui.push(moi);
    qui.sort_unstable();
    qui.dedup();
    for d in qui {
        if s.presence.en_ligne(d) {
            envoyer(s, d, json!({ "t": "ami", "ami": f }));
        }
    }
}

fn pousser_groupe(c: &Connection, s: &Social, comptes: &[i64]) {
    for &m in comptes {
        envoyer(s, m, json!({ "t": "groupe", "groupe": groupe(c, s, m) }));
    }
}

// ------------------------------------------------ arrivee et depart ---

/// Une page s'ouvre : elle recoit l'etat complet, ses amis la voient.
pub fn arrivee(moi: i64, canal: Canal) -> u64 {
    let b = base::base();
    let mut s = social();
    let t = Instant::now();
    let invitations: Vec<Value> = s.groupes.invitations(moi, t).into_iter().map(|de| fiche(&b, &s, de)).collect();
    let id = s.presence.connecter(moi, canal.clone());
    let etat = json!({
        "t": "etat",
        "moi": moi,
        "amis": liste_amis(&b, &s, moi),
        "groupe": groupe(&b, &s, moi),
        "invitations": invitations,
    });
    let _ = canal.send(etat.to_string());
    annoncer(&b, &s, moi);
    id
}

/// Une page se ferme. Rend vrai si le joueur n'a plus aucune page : ses
/// amis le voient hors ligne, et il a `GRACE` pour revenir avant de perdre
/// sa place dans son groupe (voir `grace`).
pub fn depart(moi: i64, id: u64) -> bool {
    let b = base::base();
    let mut s = social();
    let dernier = s.presence.deconnecter(moi, id);
    if dernier {
        annoncer(&b, &s, moi);
    }
    dernier
}

/// Apres le delai de grace, un joueur toujours absent quitte son groupe.
pub async fn grace(moi: i64) {
    tokio::time::sleep(GRACE).await;
    let b = base::base();
    let mut s = social();
    if s.presence.en_ligne(moi) {
        return;
    }
    let touches = s.groupes.quitter(moi);
    pousser_groupe(&b, &s, &touches);
}

// ------------------------------------------- appels du reste du site ---

/// Envoie un message a toutes les pages d'un joueur.
pub fn signaler(compte: i64, message: Value) {
    envoyer(&social(), compte, message);
}

/// La liste d'amis de `compte` a change : on la lui renvoie.
pub fn pousser_amis(compte: i64) {
    let b = base::base();
    let s = social();
    let amis = liste_amis(&b, &s, compte);
    envoyer(&s, compte, json!({ "t": "amis", "amis": amis }));
}

/// Le serveur de jeu assoit `compte` dans un salon, ou l'en fait sortir.
pub fn salon(compte: i64, code: Option<String>, prive: bool) {
    let b = base::base();
    let mut s = social();
    s.presence.salon(compte, code, prive);
    annoncer(&b, &s, compte);
}

/// Les autres membres du groupe de `compte`.
pub fn coequipiers(compte: i64) -> Vec<i64> {
    social().groupes.coequipiers(compte)
}

// ------------------------------------------------- messages recus ---

fn texte(v: &Value, cle: &str) -> String {
    v.get(cle).and_then(Value::as_str).unwrap_or("").to_string()
}

fn nombre(v: &Value, cle: &str) -> i64 {
    v.get(cle).and_then(Value::as_i64).unwrap_or(0)
}

/// Un message de la page de `moi`.
pub fn recevoir(moi: i64, brut: &str) {
    if brut.len() > 2048 {
        return;
    }
    let Ok(m) = serde_json::from_str::<Value>(brut) else { return };
    let b = base::base();
    let mut s = social();
    let t = Instant::now();
    let erreur = |s: &Social, texte: &str| envoyer(s, moi, json!({ "t": "erreur", "m": texte }));
    let nom = |c: &Connection, id: i64| joueurs::par_id(c, id).map(|j| j.pseudo).unwrap_or_default();

    match m.get("t").and_then(Value::as_str) {
        Some("lieu") => {
            s.presence.lieu(moi, &texte(&m, "lieu"));
            annoncer(&b, &s, moi);
        }
        Some("inviter") => {
            let a = nombre(&m, "a");
            if !amis::sont_amis(&b, moi, a) {
                return erreur(&s, "Tu n'invites que tes amis.");
            }
            if !s.presence.en_ligne(a) {
                return erreur(&s, &format!("{} n'est pas en ligne.", nom(&b, a)));
            }
            if let Err(e) = s.groupes.inviter(moi, a, t) {
                return erreur(&s, e);
            }
            envoyer(&s, a, json!({ "t": "invitation", "de": fiche(&b, &s, moi) }));
            envoyer(&s, moi, json!({ "t": "info", "m": format!("Invitation envoyée à {}.", nom(&b, a)) }));
        }
        Some("repondre") => {
            let de = nombre(&m, "de");
            if m.get("oui").and_then(Value::as_bool) != Some(true) {
                s.groupes.refuser(moi, de);
                envoyer(&s, de, json!({ "t": "info", "m": format!("{} a décliné ton invitation.", nom(&b, moi)) }));
                return;
            }
            match s.groupes.accepter(moi, de, t) {
                Ok(touches) => pousser_groupe(&b, &s, &touches),
                Err(e) => erreur(&s, e),
            }
        }
        Some("quitter") => {
            let touches = s.groupes.quitter(moi);
            pousser_groupe(&b, &s, &touches);
        }
        Some("exclure") => {
            let qui = nombre(&m, "qui");
            match s.groupes.exclure(moi, qui) {
                Ok(touches) => {
                    pousser_groupe(&b, &s, &touches);
                    envoyer(&s, qui, json!({ "t": "info", "m": "Tu as été exclu du groupe." }));
                }
                Err(e) => erreur(&s, e),
            }
        }
        Some("chef") => match s.groupes.passer_chef(moi, nombre(&m, "qui")) {
            Ok(touches) => pousser_groupe(&b, &s, &touches),
            Err(e) => erreur(&s, e),
        },
        // Le chef est entre dans un salon : le groupe le suit.
        Some("lancer") => {
            let code: String = texte(&m, "salon").chars().filter(char::is_ascii_digit).take(4).collect();
            if code.len() != 4 || !s.groupes.est_chef(moi) {
                return;
            }
            let chef = fiche(&b, &s, moi);
            for membre in s.groupes.coequipiers(moi) {
                envoyer(&s, membre, json!({ "t": "lancer", "salon": code, "chef": chef }));
            }
        }
        _ => {}
    }
}
